# INTERN CARRER FIRST TASK ON TITANIC DATASET

In [1]:
import pandas as pd
import numpy as np

In [2]:
df=pd.read_csv(r'C:\Users\Soumadeep\Desktop\INTERNSHIP\TITANIC\train.csv')
df

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

# EXPLORATORY DATA ANALYSIS [EDA]

In [3]:
# to calculate the total number of passengers in the dataset
df['PassengerId'].count()

891

In [4]:
# to determine the percentage of male and female passengers on board
a=list(df['Sex']).count('male')
b=list(df['Sex']).count('female')
c=df['PassengerId'].count()
male_percentage=round((a/c)*100,2)
female_percentage=round((b/c)*100,2)
print('The male percentage is:',male_percentage,'%')
print('The female percentage is:',female_percentage,'%')

The male percentage is: 64.76 %
The female percentage is: 35.24 %


In [5]:
# to find the survival rate of passengers (percentage of passengers who survived)
a=list(df['Survived']).count(1)
c=df['PassengerId'].count()
survival_percentage=round((a/c)*100,2)
print('The Survival Percentage is:',survival_percentage,"%")

The Survival Percentage is: 38.38 %


In [6]:
# to find the average age of passengers on board by Handling missing values appropriately. 
df['Age'].dropna()
print("The Average Age of Passengers on Board:",round(df['Age'].mean(),2))


The Average Age of Passengers on Board: 29.7


In [7]:
# to determine the passenger(s) with the highest fare paid
df.sort_values('Fare',ascending=False)

PassengerId  Survived  Pclass                                Name  \
258          259         1       1                    Ward, Miss. Anna   
737          738         1       1              Lesurer, Mr. Gustave J   
679          680         1       1  Cardeza, Mr. Thomas Drake Martinez   
88            89         1       1          Fortune, Miss. Mabel Helen   
27            28         0       1      Fortune, Mr. Charles Alexander   
..           ...       ...     ...                                 ...   
633          634         0       1       Parr, Mr. William Henry Marsh   
413          414         0       2      Cunningham, Mr. Alfred Fleming   
822          823         0       1     Reuchlin, Jonkheer. John George   
732          733         0       2                Knight, Mr. Robert J   
674          675         0       2          Watson, Mr. Ennis Hastings   

        Sex   Age  SibSp  Parch    Ticket      Fare        Cabin Embarked  
258  female  35.0      0      0  PC 17755  512.3292          NaN        C  
737    male  35.0      0      0  PC 17755  512.3292         B101        C  
679    male  36.0      0      1  PC 17755  512.3292  B51 B53 B55        C  
88   female  23.0      3      2     19950  263.0000  C23 C25 C27        S  
27     male  19.0      3      2     19950  263.0000  C23 C25 C27        S  
..      ...   ...    ...    ...       ...       ...          ...      ...  
633    male   NaN      0      0    112052    0.0000          NaN        S  
413    male   NaN      0      0    239853    0.0000          NaN        S  
822    male  38.0      0      0     19972    0.0000          NaN        S  
732    male   NaN      0      0    239855    0.0000          NaN        S  
674    male   NaN      0      0    239856    0.0000          NaN        S  

[891 rows x 12 columns]

# clearly the highest fare is 512.3292 & it is paid by 3 passengers whose id is 259,738,680 respectively

In [8]:
# to find the number of passengers in each passenger class (Pclass).
df.groupby('Pclass')['PassengerId'].count()

Pclass
1    216
2    184
3    491
Name: PassengerId, dtype: int64

In [9]:
# to calculate the median age of male passengers and female passengers separately.
df.groupby('Sex')['Age'].median()

Sex
female    27.0
male      29.0
Name: Age, dtype: float64

In [10]:
# to identify the passenger(s) with the longest name (based on the length of the 'Name' column).
df.sort_values('Name',ascending=False)

PassengerId  Survived  Pclass                                   Name  \
868          869         0       3            van Melkebeke, Mr. Philemon   
153          154         0       3        van Billiard, Mr. Austin Blyler   
361          362         0       2              del Carlo, Mr. Sebastiano   
282          283         0       3              de Pelsmaeker, Mr. Alfons   
286          287         1       3                de Mulder, Mr. Theodore   
..           ...       ...     ...                                    ...   
874          875         1       2  Abelson, Mrs. Samuel (Hannah Wizosky)   
308          309         0       2                    Abelson, Mr. Samuel   
279          280         1       3       Abbott, Mrs. Stanton (Rosa Hunt)   
746          747         0       3            Abbott, Mr. Rossmore Edward   
845          846         0       3                    Abbing, Mr. Anthony   

        Sex   Age  SibSp  Parch         Ticket     Fare Cabin Embarked  
868    male   NaN      0      0         345777   9.5000   NaN        S  
153    male  40.5      0      2       A/5. 851  14.5000   NaN        S  
361    male  29.0      1      0  SC/PARIS 2167  27.7208   NaN        C  
282    male  16.0      0      0         345778   9.5000   NaN        S  
286    male  30.0      0      0         345774   9.5000   NaN        S  
..      ...   ...    ...    ...            ...      ...   ...      ...  
874  female  28.0      1      0      P/PP 3381  24.0000   NaN        C  
308    male  30.0      1      0      P/PP 3381  24.0000   NaN        C  
279  female  35.0      1      1      C.A. 2673  20.2500   NaN        S  
746    male  16.0      1      1      C.A. 2673  20.2500   NaN        S  
845    male  42.0      0      0      C.A. 5547   7.5500   NaN        S  

[891 rows x 12 columns]

# A passenger whose Passenger ID is 869 has the longest Name

In [11]:
# to find the number of passengers who survived from each passenger class (Pclass).
survived_pclass_df=df[['Survived','Pclass']]
survived_pclass_df.groupby(['Pclass']).count()

Survived
Pclass          
1            216
2            184
3            491

In [12]:
# to calculate the average fare paid by passengers in each passenger class (Pclass).
fare_pclass_df=df[['Fare','Pclass']]
round(fare_pclass_df.groupby(['Pclass']).mean(),2)

Fare
Pclass       
1       84.15
2       20.66
3       13.68

In [13]:
# to find the age of the oldest and youngest passengers on board
na_dropping_df=df['Age'].dropna()
na_dropping_df.sort_values()

803     0.42
755     0.67
644     0.75
469     0.75
78      0.83
       ...  
116    70.50
493    71.00
96     71.00
851    74.00
630    80.00
Name: Age, Length: 714, dtype: float64

# THE YOUNGEST PASSENGER WAS 5 MONTHS OLD APPROAXIMATELY
# THE OLDEST PASSENGER WAS 80 YEARS OLD


In [14]:
# to identify the passengers who have siblings/spouses aboard (SibSp) and are parents/children aboard (Parch).
conditional_df=df[(df['SibSp'] ==1) & (df['Parch']!=0)]
conditional_df

PassengerId  Survived  Pclass  \
10            11         1       3   
13            14         0       3   
25            26         1       3   
43            44         1       2   
58            59         1       2   
..           ...       ...     ...   
852          853         0       3   
856          857         1       1   
869          870         1       3   
871          872         1       1   
888          889         0       3   

                                                  Name     Sex   Age  SibSp  \
10                     Sandstrom, Miss. Marguerite Rut  female   4.0      1   
13                         Andersson, Mr. Anders Johan    male  39.0      1   
25   Asplund, Mrs. Carl Oscar (Selma Augusta Emilia...  female  38.0      1   
43            Laroche, Miss. Simonne Marie Anne Andree  female   3.0      1   
58                        West, Miss. Constance Mirium  female   5.0      1   
..                                                 ...     ...   ...    ...   
852                            Boulos, Miss. Nourelain  female   9.0      1   
856         Wick, Mrs. George Dennick (Mary Hitchcock)  female  45.0      1   
869                    Johnson, Master. Harold Theodor    male   4.0      1   
871   Beckwith, Mrs. Richard Leonard (Sallie Monypeny)  female  47.0      1   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   

     Parch         Ticket      Fare Cabin Embarked  
10       1        PP 9549   16.7000    G6        S  
13       5         347082   31.2750   NaN        S  
25       5         347077   31.3875   NaN        S  
43       2  SC/Paris 2123   41.5792   NaN        C  
58       2     C.A. 34651   27.7500   NaN        S  
..     ...            ...       ...   ...      ...  
852      1           2678   15.2458   NaN        C  
856      1          36928  164.8667   NaN        S  
869      1         347742   11.1333   NaN        S  
871      1          11751   52.5542   D35        S  
888      2     W./C. 6607   23.4500   NaN        S  

[86 rows x 12 columns]

In [15]:
# to calculate the percentage of passengers who survived based on their embarked port (C = Cherbourg, Q = Queenstown, S = Southampton).
df['Survived'].isna()#THERE IS NO MISSING DATA
total_survived_df=df[(df['Survived']==1)]
total_survival=total_survived_df['Survived'].count()#THERE ARE 342 SURVIVED PASSENGERS
cherbourg_survival_df=df[(df['Survived']==1)&(df['Embarked']=='C')]
cherbourg_survival=cherbourg_survival_df['Survived'].count()#THERE ARE 93 SURVIVED PASSENGERS WHOSE EMBARKED PORT WAS CHERBOURG
queenstown_survival_df=df[(df['Survived']==1)&(df['Embarked']=='Q')]
queenstown_survival=queenstown_survival_df['Survived'].count()#THERE ARE 30 SURVIVED PASSENGERS WHOSE EMBARKED PORT WAS QUEENSTOWN
southampton_survival_df=df[(df['Survived']==1)&(df['Embarked']=='S')]
southampton_survival=southampton_survival_df['Survived'].count()#THERE ARE 217 SURVIVED PASSENGERS WHOSE EMBARKED PORT WAS SOUTHAMPTON
cherbourg_percentage=round((cherbourg_survival/total_survival)*100,2)
queenstown_percentage=round((queenstown_survival/total_survival)*100,2)
southampton_percentage=round((southampton_survival/total_survival)*100,2)
print('THE PERCENTAGE OF SURVIVED PASSENGERS FROM CHERBOURG IS',cherbourg_percentage,'%')
print('THE PERCENTAGE OF SURVIVED PASSENGERS FROM QUEENSTOWN IS',queenstown_percentage,'%')
print('THE PERCENTAGE OF SURVIVED PASSENGERS FROM SOUTHAMPTON IS',southampton_percentage,'%')


THE PERCENTAGE OF SURVIVED PASSENGERS FROM CHERBOURG IS 27.19 %
THE PERCENTAGE OF SURVIVED PASSENGERS FROM QUEENSTOWN IS 8.77 %
THE PERCENTAGE OF SURVIVED PASSENGERS FROM SOUTHAMPTON IS 63.45 %


# Model Building

In [16]:
titanic_train=pd.read_csv(r'C:\Users\Soumadeep\Desktop\INTERNSHIP\TITANIC\train.csv')
titanic_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

# For predicting the survival of any passenger the information about their name , ticket & cabin are irrelevent
# that's why we drop those columns

In [17]:
titanic_train=titanic_train.drop(['Name','Ticket','Cabin'],axis=1)
titanic_train

PassengerId  Survived  Pclass     Sex   Age  SibSp  Parch     Fare  \
0              1         0       3    male  22.0      1      0   7.2500   
1              2         1       1  female  38.0      1      0  71.2833   
2              3         1       3  female  26.0      0      0   7.9250   
3              4         1       1  female  35.0      1      0  53.1000   
4              5         0       3    male  35.0      0      0   8.0500   
..           ...       ...     ...     ...   ...    ...    ...      ...   
886          887         0       2    male  27.0      0      0  13.0000   
887          888         1       1  female  19.0      0      0  30.0000   
888          889         0       3  female   NaN      1      2  23.4500   
889          890         1       1    male  26.0      0      0  30.0000   
890          891         0       3    male  32.0      0      0   7.7500   

    Embarked  
0          S  
1          C  
2          S  
3          S  
4          S  
..       ...  
886        S  
887        S  
888        S  
889        C  
890        Q  

[891 rows x 9 columns]

# For choosing the best Machine Learning Model we use pycaret library

In [18]:
from pycaret.classification import *
s=setup(titanic_train,target='Survived')

In [19]:
best=compare_models()


Initiated  . . . . . . . . . . . . . . . . . .             08:28:07
Status     . . . . . . . . . . . . . . . . . .    Loading Estimator
Estimator  . . . . . . . . . . . . . . . . . .  Logistic Regression

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

# From the Chart above we see that Random Forest will be our Best Model for predicting as it has the Highest Accuracy i.e around 80%

In [20]:
print(best)

RandomForestClassifier(bootstrap=True, ccp_alpha=0.0, class_weight=None,
                       criterion='gini', max_depth=None, max_features='sqrt',
                       max_leaf_nodes=None, max_samples=None,
                       min_impurity_decrease=0.0, min_samples_leaf=1,
                       min_samples_split=2, min_weight_fraction_leaf=0.0,
                       n_estimators=100, n_jobs=-1, oob_score=False,
                       random_state=8153, verbose=0, warm_start=False)


# This is our Best Model for predicting

In [21]:
evaluate_model(best)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

# From feature importance plot we clearly see that some columns have don't show much importance in predicting & it is also affecting the model & it's accuracy
# That's why we drop those columns & Again use the pycaret library for detecting best machine learning model 

In [22]:
titanic_train=titanic_train.drop(['SibSp','Parch','Embarked'],axis=1)
titanic_train

PassengerId  Survived  Pclass     Sex   Age     Fare
0              1         0       3    male  22.0   7.2500
1              2         1       1  female  38.0  71.2833
2              3         1       3  female  26.0   7.9250
3              4         1       1  female  35.0  53.1000
4              5         0       3    male  35.0   8.0500
..           ...       ...     ...     ...   ...      ...
886          887         0       2    male  27.0  13.0000
887          888         1       1  female  19.0  30.0000
888          889         0       3  female   NaN  23.4500
889          890         1       1    male  26.0  30.0000
890          891         0       3    male  32.0   7.7500

[891 rows x 6 columns]

In [23]:
s=setup(titanic_train,target='Survived')

In [25]:
best_1=compare_models()

Initiated  . . . . . . . . . . . . . . . . . .              08:33:27
Status     . . . . . . . . . . . . . . . . . .  Loading Dependencies
Estimator  . . . . . . . . . . . . . . . . . .     Compiling Library

Processing:   0%|          | 0/65 [00:00<?, ?it/s]

# Now we see that the accuracy of our previous random forest model has increased but some other model came as best now that is Gradient Boosting classifier having accuracy 82% approximately

In [26]:
print(best_1)

GradientBoostingClassifier(ccp_alpha=0.0, criterion='friedman_mse', init=None,
                           learning_rate=0.1, loss='log_loss', max_depth=3,
                           max_features=None, max_leaf_nodes=None,
                           min_impurity_decrease=0.0, min_samples_leaf=1,
                           min_samples_split=2, min_weight_fraction_leaf=0.0,
                           n_estimators=100, n_iter_no_change=None,
                           random_state=762, subsample=1.0, tol=0.0001,
                           validation_fraction=0.1, verbose=0,
                           warm_start=False)


# This is our best model now

In [27]:
evaluate_model(best_1)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

# easily we can see that the features have some resonable importance for predicting the survival of the passenger
# Now we use the test data set for predicting & use that machine learning model to predict

In [28]:
titanic_test=pd.read_csv(r'C:\Users\Soumadeep\Desktop\INTERNSHIP\TITANIC\test.csv')
titanic_test

PassengerId  Pclass                                          Name  \
0            892       3                              Kelly, Mr. James   
1            893       3              Wilkes, Mrs. James (Ellen Needs)   
2            894       2                     Myles, Mr. Thomas Francis   
3            895       3                              Wirz, Mr. Albert   
4            896       3  Hirvonen, Mrs. Alexander (Helga E Lindqvist)   
..           ...     ...                                           ...   
413         1305       3                            Spector, Mr. Woolf   
414         1306       1                  Oliva y Ocana, Dona. Fermina   
415         1307       3                  Saether, Mr. Simon Sivertsen   
416         1308       3                           Ware, Mr. Frederick   
417         1309       3                      Peter, Master. Michael J   

        Sex   Age  SibSp  Parch              Ticket      Fare Cabin Embarked  
0      male  34.5      0      0              330911    7.8292   NaN        Q  
1    female  47.0      1      0              363272    7.0000   NaN        S  
2      male  62.0      0      0              240276    9.6875   NaN        Q  
3      male  27.0      0      0              315154    8.6625   NaN        S  
4    female  22.0      1      1             3101298   12.2875   NaN        S  
..      ...   ...    ...    ...                 ...       ...   ...      ...  
413    male   NaN      0      0           A.5. 3236    8.0500   NaN        S  
414  female  39.0      0      0            PC 17758  108.9000  C105        C  
415    male  38.5      0      0  SOTON/O.Q. 3101262    7.2500   NaN        S  
416    male   NaN      0      0              359309    8.0500   NaN        S  
417    male   NaN      1      1                2668   22.3583   NaN        C  

[418 rows x 11 columns]

# we also drop those irrelevant columns from the test dataset also

In [29]:
titanic_test=titanic_test.drop(['Name','Ticket','Cabin','SibSp','Parch','Embarked'],axis=1)
titanic_test

PassengerId  Pclass     Sex   Age      Fare
0            892       3    male  34.5    7.8292
1            893       3  female  47.0    7.0000
2            894       2    male  62.0    9.6875
3            895       3    male  27.0    8.6625
4            896       3  female  22.0   12.2875
..           ...     ...     ...   ...       ...
413         1305       3    male   NaN    8.0500
414         1306       1  female  39.0  108.9000
415         1307       3    male  38.5    7.2500
416         1308       3    male   NaN    8.0500
417         1309       3    male   NaN   22.3583

[418 rows x 5 columns]

In [30]:
titanic_test=predict_model(best_1,titanic_test)
titanic_test

PassengerId  Pclass     Sex   Age        Fare  prediction_label  \
0            892       3    male  34.5    7.829200                 0   
1            893       3  female  47.0    7.000000                 0   
2            894       2    male  62.0    9.687500                 0   
3            895       3    male  27.0    8.662500                 0   
4            896       3  female  22.0   12.287500                 0   
..           ...     ...     ...   ...         ...               ...   
413         1305       3    male   NaN    8.050000                 0   
414         1306       1  female  39.0  108.900002                 1   
415         1307       3    male  38.5    7.250000                 0   
416         1308       3    male   NaN    8.050000                 0   
417         1309       3    male   NaN   22.358299                 0   

     prediction_score  
0              0.9374  
1              0.9393  
2              0.9371  
3              0.8784  
4              0.5193  
..                ...  
413            0.8913  
414            0.9654  
415            0.9707  
416            0.8913  
417            0.8267  

[418 rows x 7 columns]

# the column prediction_label is our prediction of survival here